### PR TITLE
LockWatcher: Keep an eye on Windows file locks

### PR DESCRIPTION
**Pull request by ckamm**
When a conflict-rename or a temporary-rename fails, notify the
LockWatcher. It'll regularly check whether the file has become
accesible again. When it has, another sync is triggered.

owncloud/enterprise#1288

### CODE REVIEW

File: src/gui/folderman.cpp [hunk 1]
```diff
@@ -744,6 +754,16 @@ void FolderMan::slotServerVersionChanged(Account *account)
     }
 }
 
+void FolderMan::slotScheduleFolderOwningFile(const QString& path)
+{
+    QString cleanFilePath = QDir::cleanPath(path);
+    foreach (Folder* f, _folderMap) {
+        if (cleanFilePath.startsWith(f->cleanPath())) {
+            slotScheduleSync(f);
```
Comment (guruz) [inline src/gui/folderman.cpp]: Folder* FolderMan::folderForPath(const QString& path);

File: src/gui/folderman.h [hunk 2]
```diff
@@ -227,6 +241,7 @@ private slots:
     QPointer<RequestEtagJob>        _currentEtagJob; // alias of Folder running the current RequestEtagJob
 
     QMap<QString, FolderWatcher*> _folderWatchers;
+    LockWatcher* _lockWatcher;
```
Comment (guruz) [inline src/gui/folderman.h]: @ogoffart Would you say it's prefered to use QScopedPointer?

File: src/gui/lockwatcher.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+#include "lockwatcher.h"
+#include "filesystem.h"
+
+#include <QTimer>
+
+using namespace OCC;
+
+static const int check_frequency = 20000; // ms
```
Comment (guruz) [inline src/gui/lockwatcher.cpp]: nitpick: Could write this as 20*1000

File: src/gui/lockwatcher.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    /** Emitted when one of the watched files is no longer
+     *  being locked. */
+    void fileUnlocked(const QString& path);
+
+private slots:
+    void checkFiles();
+
+private:
+    QSet<QString> _watchedPaths;
+    QTimer* _timer;
```
Comment (guruz) [inline src/gui/lockwatcher.h]: Add as normal member, not pointer?

File: src/gui/folderman.h (hunk 2 above)
Comment (ogoffart) [inline src/gui/folderman.h]: yes.  

QScopedPointer document the ownership.

Comment (guruz) [inline src/gui/folderman.h]: ah wait, why not have it as a normal member?

File: src/gui/lockwatcher.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) by Klaas Freitag <freitag@owncloud.com>
```
Comment (dragotin) [inline src/gui/lockwatcher.cpp]: nitpick: My copyright??

File: src/gui/lockwatcher.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <QTimer>
+
+using namespace OCC;
+
+static const int check_frequency = 20000; // ms
+
+LockWatcher::LockWatcher(QObject* parent)
+    : QObject(parent)
+    , _timer(new QTimer(this))
```
Comment (dragotin) [inline src/gui/lockwatcher.cpp]: Any reason to not have _timer as a normal member rather than allocating?

File: src/gui/lockwatcher.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#ifndef MIRALL_LOCKWATCHER_H
+#define MIRALL_LOCKWATCHER_H
```
Comment (dragotin) [inline src/gui/lockwatcher.h]: Please do not use MIRALL_ namespace.

**Comment by ckamm** [comment]
@dragotin @guruz Thanks for the review, those comments are addressed. I'll only be back in two weeks, so feel free to fix and merge anything else that comes up during testing!

**Comment by guruz** [comment]
+1 Looks good to me, but somebody else needs to approve

**Comment by dragotin** [comment]
Tested on windows with Word and conflicts, works nicely. Awesome :+1:

**Pull request merged**